### ISSUE TITLE
Static binning solution for voroni

### ISSUE THREAD

**Issue by eanderson4** [ResidentMario/geoplot#182]
I'm looking to create static category bins so my coloring doesn't change between plots.  I'm not sure how to achieve this sort of behavior currently and was looking for assistance.

My voroni call looks like this, and the data is changing between invocations
```
gplt.voronoi(
		    data, 
		    projection=ep.proj, clip=clipper,
		    hue='lmp', cmap='Reds', 
		    k=10, scheme='fisher_jenks',
		    ax=ax, alpha=0.9,zorder=7,
		    edgecolor='white', legend=True,extent=clipper.total_bounds
		    )
```

ex. plot 1
![image](https://user-images.githubusercontent.com/5291870/67796033-e9476080-fa44-11e9-9e64-68fa12957488.png)

ex. plot 2
![image](https://user-images.githubusercontent.com/5291870/67796038-eba9ba80-fa44-11e9-9e74-ac4887fdb7dd.png)

I would just like to create the bins beforehand, or maybe preprocess the data in some way so that the legend and colors are consistent between plots so I can use them in an animation.

Thanks!

**Comment by ResidentMario** [comment]
Thanks for reporting this issue!

You are not the first person to ask for this "colormap normalization between plots" feature; another user asked for this a few months ago in #163.

The TLDR is that to reuse the same colormap across plots, you need to add a normalization function to your plotter (via the `norm` parameter). In that user's case the normalization function needed was very simple: `norm=colors.Normalize(vmin=$BOTTOM_VALUE, vmax=$TOP_VALUE)`. In your case however, since you are using a categorical colormap, the recipe needed is a lot more complicated. You will need to roll your own custom normalization class. The following code snippet demonstrates:

```python
import geoplot as gplt
import pandas as pd
import geopandas as gpd

injurious_collisions = gpd.read_file(gplt.datasets.get_path('nyc_injurious_collisions'))

from matplotlib import colors
from pysal.esda.mapclassify import Quantiles

class Groupwise_Quantiles(colors.Normalize):
    def __init__(self, y, k=5, vmin=None, vmax=None, clip=False):
        # your classifier goes here!
        self.clf = Quantiles(y, k=k)
        self.vmin, self.vmax = 0, len(self.clf.bins) - 1

    def __call__(self, value, clip=None):
        # at call time, transform into the [0, 1] range
        return self.clf(value) / self.vmax


fig, axarr = plt.subplots(1, 2, figsize=(16, 6))

norm = Groupwise_Quantiles(injurious_collisions['NUMBER OF PERSONS INJURED'])

gplt.pointplot(
    injurious_collisions[injurious_collisions['NUMBER OF PERSONS INJURED'] <= 5],
    hue='NUMBER OF PERSONS INJURED',
    cmap='viridis',
    norm=norm,
    ax=axarr[0]
)
gplt.pointplot(
    injurious_collisions[injurious_collisions['NUMBER OF PERSONS INJURED'] >= 3],
    hue='NUMBER OF PERSONS INJURED',
    cmap='viridis',
    norm=norm,
    ax=axarr[1]
)
```

You should be able to adapt this code sample to your own use case.

----

When I originally designed this library, the "many plots, same colormap" use-case did not occur to me, and I considered it an advanced use case that I wasn't particularly going to optimize for. However, the volume of recent questions regarding this feature demonstrates otherwise.

There have also been an underlying change to the API that `geoplot` relies on for the implementation of this feature: [`pysal.edsa`](https://pysal.readthedocs.io/en/v1.11.0/library/esda/mapclassify.html#pysal.esda.mapclassify.Quantiles) is now functional instead of string-based.

In the next major update I'm going to remove the `k` parameter from the API and allow `mapclassify.edsa` classes (and custom user-written functors) as input to `scheme`. This should make reusing the same classification scheme across multiple plots super-easy:

```python
scheme = <pysal.esda.mapclassify.Quantiles object>

# different plots, same classification scheme
gplt.pointplot(hue='foo1', cmap='viridis', scheme=scheme)
gplt.pointplot(hue='foo2', cmap='viridis', scheme=scheme)
```

**Comment by eanderson4** [comment]
Thanks for the quick response and thorough response!  Sorry I missed the discussion on this issue when looking around.  

Your approach looks like it will work well for this, I'll give a quick update when I finish implementing.

**Comment by eanderson4** [comment]
I'm not quite sure I was able to achieve the desired solution.  It was hard for me to tell what was going on and I don't have a lot of time to trouble shoot for now.

I first tried using a UserDefined binning scheme from mapclassify.  In order to pass through the geopandas mapclassify, it allows kwargs to be passed through for the scheme, so I just modified that part of your library to pass through the bins kwarg for the userdefined scheme.  However, once I finally got it working, the bins were not well designed and it seems using a good binning scheme is helpful.

So I went back to your example and just changed to the FisherJenks scheme and initialized it with the values from all my plots, not just a single plot.  It seems like it is still changing between plots and for sure the legend is changing. I couldn't tell if this was a bug with the legend and norm functionality or it was just my norm scheme not working how I want it to.

I did print the values in the call() function for the GroupwiseQuantile function and it seems like it was being called with values that were already binned, i.e. they were all small integers.  Is that what is suppose to be happening?

Anyways, here was the norm function, voroni call, and my outputing plots.  I'm closing for now as I'm just going to deal with changing scales for now and when I have more time I might attempt again.

```

from matplotlib import colors
from mapclassify import FisherJenks

class Groupwise_Quantiles(colors.Normalize):
    def __init__(self, y, k=5, vmin=None, vmax=None, clip=False):
        # your classifier goes here!
        self.clf = FisherJenks(y, k=k)
        self.vmin, self.vmax = 0, len(self.clf.bins) - 1

    def __call__(self, value, clip=None):
        # at call time, transform into the [0, 1] range
        v= value/self.vmax
        return v
norm=Groupwise_Quantiles(lmp_values,k=12)
```
```
		gplt.voronoi(
			data,  // This is a subset of the data in the lmp_values sent to the norm constructor
			projection=ep.proj, clip=clipper,
			hue='lmp', cmap='Reds', scheme='fisherjenks',k=12,
			legend=True,
			ax=ax, alpha=0.9,zorder=7,
			norm=norm,
			edgecolor='white',extent=clipper.total_bounds
			)
```

![aaLMP_Plot_2021-12-27_15:00:00](https://user-images.githubusercontent.com/5291870/67828678-f6426f00-fa99-11e9-9530-041f0c02f50e.png)
![aaLMP_Plot_2021-12-27_14:00:00](https://user-images.githubusercontent.com/5291870/67828680-f6426f00-fa99-11e9-9989-92e3f365acc9.png)

**Comment by ResidentMario** [comment]
See #184.

**Comment by eanderson4** [comment]
That will open up a lot of flexibility, looking forward to it.

I just realized I probably could have finished the workaround hack as such.

1. Create Groupwise_Quantiles as above.  I believe you can get the bins out of the classifier.
2. Pass through the scheme_kwargs bin to the geopandas mapclassifier function using the bins from the above classifier.

Perhaps I'll get to testing it out later this week.  Eitherway, as you pointed out, it definitely feels more hacked together then being able to provide a scheme directly.

One thought I had on this would be to potentially allow for a pass through of scheme kwargs, so we could do things such as what I'm trying to do above, like

```
scheme='userdefined'
scheme_kwargs={ 'bins' :[0,10,20,30,40] }
```

The scheme_kwargs would be passed through to geopandas mapclassifier and it would be some pretty expressive and functional api.

Just some thoughts.

Thanks again, it would have taken ages to get reasonable plots for my project and they wouldn't have looked as good!

**Comment by ResidentMario** [comment]
> Thanks again, it would have taken ages to get reasonable plots for my project and they wouldn't have looked as good!

Thank you for the kind words.

Would it be possible for you to share the data and/or code that you are using to generate these plots? I want to add an example to the gallery demonstrating categorization scheme reuse. It would be great to have a fresh dataset to demonstrate it on, and it looks like you've already done much of the necessary work. 🙂

**Comment by eanderson4** [comment]
Yah, I can do that. Give me a few days so I can make sure there is no confidential info and I'll set up a repo with the data and the test plots.

**Comment by ResidentMario** [comment]
Great! That would be super helpful. Thanks.

**Comment by ResidentMario** [comment]
The `scheme` update has been merged and a new 0.4.0 release has been minted with this change (live on PyPI, coming soon to conda-forge). This should be very easy to do now. A couple of references:

* https://github.com/ResidentMario/geoplot/releases/tag/0.4.0
* https://residentmario.github.io/geoplot/gallery/plot_california_districts.html

**Comment by eanderson4** [comment]
That is awesome!
